Digipost.Signature.Api.Client.Core/Internal/LoggingHandler.cs

```diff
@@ -1,28 +1,95 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
 using System.Threading;
 using System.Threading.Tasks;
+using log4net;
 
 namespace Digipost.Signature.Api.Client.Core.Internal
 {
     internal class LoggingHandler : DelegatingHandler
     {
-        public LoggingHandler(HttpMessageHandler innerHandler)
-            : base(innerHandler)
-        {
-        }
+        private static readonly ILog Log = LogManager.GetLogger("Digipost.Signature.Api.Client.RequestLogger");
 
         protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
             CancellationToken cancellationToken)
         {
-            //Todo: Add logging before sending
+            if (Log.IsDebugEnabled)
+            {
+                await LogRequest(request);
+            }
+
+            var httpResponseMessage = await base.SendAsync(request, cancellationToken);
+
+            if (Log.IsDebugEnabled)
+            {
+                await LogResponse(httpResponseMessage);
+            }
+
+            return httpResponseMessage;
+        }
+
+        private static async Task LogRequest(HttpRequestMessage request)
+        {
+            var requestData = await GetRequestData(request);
+            Log.Debug($"Outgoing: {requestData}");
+        }
+
+        private static async Task<string> GetRequestData(HttpRequestMessage request)
+        {
+            var keyValuePairs = new List<KeyValuePair<string, string>>
+            {
+                new KeyValuePair<string, string>("Method", $"{request.Method}, {request.RequestUri}"),
+                new KeyValuePair<string, string>("Accept", $"{request.Headers.Accept}"),
+                new KeyValuePair<string, string>("UserAgent", $"{request.Headers.UserAgent}"),
+                new KeyValuePair<string, string>(null, $"{await GetContentData(request.Content)}")
+            };
+
+            return FormatHttpData(keyValuePairs);
+        }
+
+        private static async Task LogResponse(HttpResponseMessage response)
+        {
+            var responseData = await GetResponseData(response);
+            Log.Debug($"Incoming: {responseData}");
+        }
 
-            var response = await base.SendAsync(request, cancellationToken);
+        private static async Task<string> GetResponseData(HttpResponseMessage response)
+        {
+            var keyValuePairs = new List<KeyValuePair<string, string>>
+            {
+                new KeyValuePair<string, string>("StatusCode", $"{(int) response.StatusCode}, {response.StatusCode}"),
+                new KeyValuePair<string, string>("Cache-Control", $"{response.Headers.CacheControl}"),
+                new KeyValuePair<string, string>("Content-Length", $"{response.Content.Headers.ContentLength}"),
+                new KeyValuePair<string, string>("Content-Type", $"{response.Content.Headers.ContentType}"),
+                new KeyValuePair<string, string>("Date", $"{response.Headers.Date}"),
```

Comment:
Would it be possible to iterate over and print *all* the headers rather than pre defining certain values?
Some responses from the server side might have other headers than these (e.g. the `X-Next-permitted-poll-time` returned when polling for status changes), and it would be useful to log everything in case we choose to add other headers in other responses in the future.

The same applies for when logging the request :)